core/src/main/java/org/apache/calcite/rel/type/RelDataTypeSystem.java

```diff
@@ -297,6 +297,7 @@ default boolean shouldUseDoubleMultiplication(RelDataTypeFactory typeFactory,
         int six = Math.min(6, getMaxNumericScale());
         int d = p1 - s1 + s2;
         int scale = Math.max(six, s1 + p2 + 1);
+        scale = Math.min(scale, getMaxNumericScale());
```

Comment:
I don't quite understand the rules, but I see some paradoxes.
Assume the following scenario.
```
six = Math.min(6, getMaxNumericScale()) = getMaxNumericScale()
scale = Math.max(six, s1 + p2 + 1) = six = getMaxNumericScale()
scale = Math.min(scale, getMaxNumericScale()) = getMaxNumericScale()
```
Is this a repeated judgment?